File: lib/octocon_web/controllers/auth_link_controller.ex
defmodule OctoconWeb.AuthLinkController do
  use OctoconWeb, :controller

  require Logger

  alias Octocon.Global.LinkTokenRegistry

  plug :put_link_token
  plug Ueberauth, providers: [:google_link, :discord_link]

  def request(conn, _params) do
    conn
    |> send_resp(403, "")
  end

  def put_link_token(%{params: %{"link_token" => link_token}} = conn, _params) do
    Logger.info("Link token injected into session: #{link_token}")

    conn
    |> put_session(:link_token, link_token)
  end

  def put_link_token(conn, _params) do
    conn
  end

  def callback(%{assigns: %{ueberauth_auth: auth_data}} = conn, %{"provider" => provider}) do
    link_token = get_session(conn, :link_token)
    Logger.info("Link token received: #{link_token}")

    case Fly.RPC.rpc_primary(fn -> LinkTokenRegistry.get(link_token) end) do
      nil ->
        conn
        |> delete_session(:link_token)
        |> put_status(403)
        |> text("This link token is invalid or has expired.")

      system_id when is_binary(system_id) ->
        Logger.info("Recovered ID: #{system_id}")
        LinkTokenRegistry.delete(link_token)

        try_link_account_wrapper(provider, conn, system_id, auth_data)
    end
  end

  def callback(conn, _params) do
    conn
    |> put_status(403)
    |> text("Failed to authenticate. Did you reload the page or copy-paste the URL?")
  end

  def try_link_account_wrapper(provider, conn, system_id, auth_data) do
    case Octocon.Accounts.get_user({:system, system_id}) do
      nil ->
        conn
        |> put_status(403)
        |> json(%{error: "System not found"})

      user ->
        try_link_account(provider, conn, user, auth_data)
    end
  end

  def try_link_account("discord", conn, user, %{uid: discord_id}) do
    case Octocon.Accounts.link_discord_to_user(user, to_string(discord_id)) do
      {:error, :already_linked} ->
        conn
        |> put_status(403)
        |> json(%{
          error: "A Discord account is already linked to this account; please unlink it first."
        })

      {:error,
       %Ecto.Changeset{
         errors: [
           discord_id: {"has already been taken", _}
         ]
       }} ->
        conn
        |> put_status(500)
        |> json(%{error: "This Discord account is already linked to another account."})

      {:ok, _} ->
        conn
        |> redirect(external: "https://octocon.app/deep/link_success/discord")
    end
  end

  def try_link_account("google", conn, user, %{info: %{email: email}}) do
    case Octocon.Accounts.link_email_to_user(user, email) do
      {:error, :already_linked} ->
        conn
        |> put_status(403)
        |> json(%{
          error: "A Google account is already linked to this account; please unlink it first."
        })

      {:error,
       %Ecto.Changeset{
         errors: [
           email: {"has already been taken", _}
         ]
       }} ->
        conn
        |> put_status(500)
        |> json(%{error: "This email address is already linked to another account."})

      {:ok, _} ->
        conn
        |> redirect(external: "https://octocon.app/deep/link_success/google")
    end
  end
end

File: apis/Interfold.Auth.Api/Controllers/AuthLinkController.cs
using Interfold.Auth.Api.Auth;
using Interfold.Auth.Api.Controllers.Base;
using Interfold.Auth.Api.Models;
using Interfold.Auth.Api.Services;
using Interfold.Auth.Contracts.Configuration;
using Interfold.Auth.Contracts.Enums;
using Interfold.Auth.Contracts.Models.Commands;
using Interfold.Auth.Domain;
using Interfold.Shared.Api.Models;
using Interfold.Shared.Contracts;
using Interfold.Shared.Contracts.Enums;
using Interfold.Shared.Contracts.Ids;
using Interfold.Shared.Contracts.Models.Read;
using Interfold.Shared.Contracts.Operations;
using Microsoft.AspNetCore.Authentication;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;

namespace Interfold.Auth.Api.Controllers;

[AllowAnonymous]
[Route("auth/link")]
public sealed class AuthLinkController : OAuthControllerBase
{
    private const string LinkTokenCookieName = InterfoldCookieNames.LinkToken;
    private const string RedirectUriCookieName = InterfoldCookieNames.LinkRedirectUri;

    private readonly LinkOAuthIdentityCommandHandler _linkHandler;

    public AuthLinkController(
        LinkOAuthIdentityCommandHandler linkHandler,
        IOptionsMonitor<AuthenticationConfiguration> authOptions,
        IAuthenticationSchemeProvider schemeProvider,
        GoogleOAuthService googleOAuth,
        DiscordOAuthService discordOAuth,
        AppleOAuthService appleOAuth)
        : base(authOptions, schemeProvider, googleOAuth, discordOAuth, appleOAuth)
    {
        _linkHandler = linkHandler;
    }

    protected override string CallbackRoutePrefix => "auth/link";

    [HttpGet("{provider}")]
    public async Task<IActionResult> Begin([FromRoute] string provider)
    {
        if (!provider.TryParseWire<OAuthProvider>(out var oauthProvider))
            return UnsupportedProviderResponse(provider);

        // Fail-fast: the client is contractually required to pass redirect_uri on Begin so
        // the callback (which reads it back from the octocon_link_redirect_uri cookie) can
        // route the user home after the OAuth round-trip. Detecting it upfront avoids the
        // full challenge round-trip landing in the 400 branch of RedirectToClient — the
        // client sees the same error code, just before the browser leaves for the provider.
        // Sibling AuthController.Begin has the same contract; the check lives here as a
        // per-endpoint guard rather than base-class middleware because only the Begin shape
        // carries the redirect_uri query param.
        var redirectUri = Request.Query[OAuthQueryKeys.RedirectUri].ToString();
        if (string.IsNullOrWhiteSpace(redirectUri))
        {
            Response.Headers[InterfoldHeaders.OperationId] = OperationIds.QueryAuthLinkRequest.Value;
            return BadRequest(new ErrorResponse(
                "Missing client-supplied redirect_uri.",
                ErrorCodes.MissingRedirectUri,
                detail: "Pass redirect_uri on GET /auth/link/{provider} so the link callback knows where to send the result."));
        }

        StoreQueryCookie(LinkTokenCookieName, OAuthQueryKeys.LinkToken);
        StoreRedirectUriCookie(RedirectUriCookieName);

        var challenge = await IssueChallengeIfRegisteredAsync(oauthProvider, OperationIds.QueryAuthLinkRequest);

        if (challenge is not null)
            return challenge;

        Response.Headers[InterfoldHeaders.OperationId] = OperationIds.QueryAuthLinkRequest.Value;
        return StatusCode(StatusCodes.Status403Forbidden, string.Empty);
    }

    [HttpGet("{provider}/callback")]
    public Task<IActionResult> CallbackGet([FromRoute] string provider)
        => Callback(provider);

    [HttpPost("{provider}/callback")]
    public Task<IActionResult> CallbackPost([FromRoute] string provider)
        => Callback(provider);

    private async Task<IActionResult> Callback(string provider)
    {
        if (!provider.TryParseWire<OAuthProvider>(out var oauthProvider))
            return UnsupportedProviderResponse(provider);

        // LinkToken.From wraps the query-or-cookie fallback in one call so the null check
        // operates on the typed LinkToken?, not on a bare string local. The null-coalescing
        // chain preserves its shape (query first, cookie second, .From consumes the
        // possibly-null result). A bare string local would stay alive across the null-check
        // / error-return boundary; a redirect-log or exception-message-containing-locals in
        // that window would leak the token verbatim.
        var linkToken = LinkToken.From(await GetValueAsync(OAuthQueryKeys.LinkToken) ?? Request.Cookies[LinkTokenCookieName]);
        if (linkToken is null)
        {
            return StatusCode(StatusCodes.Status403Forbidden, "This link token is invalid or has expired.");
        }

        var identity = await ExtractProviderIdentityAsync(oauthProvider);
        if (identity is not { } typedIdentity)
        {
            return StatusCode(StatusCodes.Status403Forbidden, "Failed to authenticate. Did you reload the page or copy-paste the URL?");
        }

        // [AllowAnonymous] endpoint: the middleware doesn't populate the principal, but
        // BuildEnvelope stamps the synthetic AnonymousPrincipalId in that case. The link
        // handler resolves the real system id off the link token itself
        // (ResolveSystemIdByLinkTokenAsync), so command.PrincipalId is never read.
        var envelope = BuildEnvelope(OperationIds.AuthLinkCallback, new LinkOAuthIdentityCommand(linkToken.Value, typedIdentity));
        var commandResult = await _linkHandler.HandleAsync(envelope, HttpContext.RequestAborted);
        if (!commandResult.Accepted || commandResult.Result is null)
        {
            Response.Cookies.Delete(LinkTokenCookieName);
            return StatusCode(StatusCodes.Status403Forbidden, "This link token is invalid or has expired.");
        }

        var result = commandResult.Result.Result;
        var systemId = commandResult.Result.SystemId;

        Response.Cookies.Delete(LinkTokenCookieName);

        var redirectUri = Request.Cookies[RedirectUriCookieName];
        Response.Cookies.Delete(RedirectUriCookieName);

        Response.Headers[InterfoldHeaders.OperationId] = OperationIds.AuthLinkCallback.Value;

        return result switch
        {
            AccountLinkResult.Success when systemId != null => RedirectToClient(redirectUri),
            AccountLinkResult.AlreadyLinked => StatusCode(StatusCodes.Status403Forbidden, new ErrorMessageResponse(
                oauthProvider switch
                {
                    OAuthProvider.Discord => "A Discord account is already linked to this account; please unlink it first.",
                    OAuthProvider.Google => "A Google account is already linked to this account; please unlink it first.",
                    OAuthProvider.Apple => "An Apple account is already linked to this account; please unlink it first.",
                    _ => "An account is already linked to this account; please unlink it first."
                })),
            AccountLinkResult.UserExists => StatusCode(StatusCodes.Status500InternalServerError, new ErrorMessageResponse(
                oauthProvider switch
                {
                    OAuthProvider.Discord => "This Discord account is already linked to another account.",
                    OAuthProvider.Google => "This email address is already linked to another account.",
                    OAuthProvider.Apple => "This Apple account is already linked to another account.",
                    _ => "This account is already linked to another account."
                })),
            _ => StatusCode(StatusCodes.Status403Forbidden, new ErrorMessageResponse("System not found"))
        };
    }

    private IActionResult RedirectToClient(string? redirectUri)
    {
        // The client is responsible for supplying its own redirect_uri on the initial
        // GET /auth/link/{provider}?redirect_uri=... call; the cookie threads it through
        // to here. Absence means the client never set it — surface that loudly rather
        // than papering over with a server-configured fallback.
        if (string.IsNullOrWhiteSpace(redirectUri))
        {
            return BadRequest(new ErrorResponse(
                "Missing client-supplied redirect_uri.",
                ErrorCodes.MissingRedirectUri,
                detail: "Pass redirect_uri on GET /auth/link/{provider} so the link callback knows where to send the result."));
        }

        return Redirect(redirectUri);
    }
}
